// ml-peak-shape-generator/lib/shapes/1d/lorentzianDispersive/LorentzianDispersive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLorentzianDispersiveData = exports.lorentzianDispersiveFct = exports.LorentzianDispersive = void 0;
const Lorentzian_1 = require("../lorentzian/Lorentzian");
class LorentzianDispersive {
    constructor(options = {}) {
        const { fwhm = 500 } = options;
        this.fwhm = fwhm;
    }
    fwhmToWidth(fwhm = this.fwhm) {
        return (0, Lorentzian_1.lorentzianFwhmToWidth)(fwhm);
    }
    widthToFWHM(width) {
        return (0, Lorentzian_1.lorentzianWidthToFWHM)(width);
    }
    fct(x) {
        return (0, exports.lorentzianDispersiveFct)(x, this.fwhm);
    }
    //eslint-disable-next-line
    getArea(_height) {
        return 0;
    }
    getFactor(area) {
        return (0, Lorentzian_1.getLorentzianFactor)(area);
    }
    getData(options = {}) {
        return (0, exports.getLorentzianDispersiveData)(this, options);
    }
    calculateHeight(area = 1) {
        return (0, Lorentzian_1.calculateLorentzianHeight)({ fwhm: this.fwhm, area });
    }
    getParameters() {
        return ['fwhm'];
    }
}
exports.LorentzianDispersive = LorentzianDispersive;
const lorentzianDispersiveFct = (x, fwhm) => {
    return (2 * fwhm * x) / (4 * x ** 2 + fwhm ** 2);
};
exports.lorentzianDispersiveFct = lorentzianDispersiveFct;
const getLorentzianDispersiveData = (shape = {}, options = {}) => {
    let { fwhm = 500 } = shape;
    let { length, factor = (0, Lorentzian_1.getLorentzianFactor)(), height = (0, Lorentzian_1.calculateLorentzianHeight)({ fwhm, area: 1 }), } = options;
    if (!length) {
        length = Math.min(Math.ceil(fwhm * factor), Math.pow(2, 25) - 1);
        if (length % 2 === 0)
            length++;
    }
    const center = (length - 1) / 2;
    const data = new Float64Array(length);
    for (let i = 0; i <= center; i++) {
        data[i] = (0, exports.lorentzianDispersiveFct)(i - center, fwhm) * height;
        data[length - 1 - i] = -data[i];
    }
    return data;
};
exports.getLorentzianDispersiveData = getLorentzianDispersiveData;
