// ml-peak-shape-generator/lib/shapes/1d/lorentzian/Lorentzian.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLorentzianData = exports.getLorentzianFactor = exports.lorentzianFwhmToWidth = exports.lorentzianWidthToFWHM = exports.lorentzianFct = exports.getLorentzianArea = exports.calculateLorentzianHeight = exports.Lorentzian = void 0;
const constants_1 = require("../../../util/constants");
class Lorentzian {
    constructor(options = {}) {
        const { fwhm = 500 } = options;
        this.fwhm = fwhm;
    }
    fwhmToWidth(fwhm = this.fwhm) {
        return (0, exports.lorentzianFwhmToWidth)(fwhm);
    }
    widthToFWHM(width) {
        return (0, exports.lorentzianWidthToFWHM)(width);
    }
    fct(x) {
        return (0, exports.lorentzianFct)(x, this.fwhm);
    }
    getArea(height = 1) {
        return (0, exports.getLorentzianArea)({ fwhm: this.fwhm, height });
    }
    getFactor(area) {
        return (0, exports.getLorentzianFactor)(area);
    }
    getData(options = {}) {
        return (0, exports.getLorentzianData)(this, options);
    }
    calculateHeight(area = 1) {
        return (0, exports.calculateLorentzianHeight)({ fwhm: this.fwhm, area });
    }
    getParameters() {
        return ['fwhm'];
    }
}
exports.Lorentzian = Lorentzian;
const calculateLorentzianHeight = ({ fwhm = 1, area = 1 }) => {
    return (2 * area) / Math.PI / fwhm;
};
exports.calculateLorentzianHeight = calculateLorentzianHeight;
const getLorentzianArea = (options) => {
    const { fwhm = 500, height = 1 } = options;
    return (height * Math.PI * fwhm) / 2;
};
exports.getLorentzianArea = getLorentzianArea;
const lorentzianFct = (x, fwhm) => {
    return fwhm ** 2 / (4 * x ** 2 + fwhm ** 2);
};
exports.lorentzianFct = lorentzianFct;
const lorentzianWidthToFWHM = (width) => {
    return width * constants_1.ROOT_THREE;
};
exports.lorentzianWidthToFWHM = lorentzianWidthToFWHM;
const lorentzianFwhmToWidth = (fwhm) => {
    return fwhm / constants_1.ROOT_THREE;
};
exports.lorentzianFwhmToWidth = lorentzianFwhmToWidth;
const getLorentzianFactor = (area = 0.9999) => {
    if (area >= 1) {
        throw new Error('area should be (0 - 1)');
    }
    const halfResidual = (1 - area) * 0.5;
    const quantileFunction = (p) => Math.tan(Math.PI * (p - 0.5));
    return ((quantileFunction(1 - halfResidual) - quantileFunction(halfResidual)) / 2);
};
exports.getLorentzianFactor = getLorentzianFactor;
const getLorentzianData = (shape = {}, options = {}) => {
    let { fwhm = 500 } = shape;
    let { length, factor = (0, exports.getLorentzianFactor)(), height = (0, exports.calculateLorentzianHeight)({ fwhm, area: 1 }), } = options;
    if (!length) {
        length = Math.min(Math.ceil(fwhm * factor), Math.pow(2, 25) - 1);
        if (length % 2 === 0)
            length++;
    }
    const center = (length - 1) / 2;
    const data = new Float64Array(length);
    for (let i = 0; i <= center; i++) {
        data[i] = (0, exports.lorentzianFct)(i - center, fwhm) * height;
        data[length - 1 - i] = data[i];
    }
    return data;
};
exports.getLorentzianData = getLorentzianData;
