// ml-spectra-fitting/lib/util/selectMethod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectMethod = selectMethod;
const ml_levenberg_marquardt_1 = require("ml-levenberg-marquardt");
const directOptimization_1 = require("./wrappers/directOptimization");
/** Algorithm to select the method.
 * @param optimizationOptions - Optimization options
 * @returns - The algorithm and optimization options
 */
function selectMethod(optimizationOptions = {}) {
    const { kind = 'lm', options } = optimizationOptions;
    switch (kind) {
        case 'lm':
        case 'levenbergMarquardt':
            return {
                algorithm: ml_levenberg_marquardt_1.levenbergMarquardt,
                optimizationOptions: {
                    damping: 1.5,
                    maxIterations: 100,
                    errorTolerance: 1e-8,
                    ...options,
                },
            };
        case 'direct': {
            return {
                algorithm: directOptimization_1.directOptimization,
                optimizationOptions: {
                    iterations: 20,
                    epsilon: 1e-4,
                    tolerance: 1e-16,
                    tolerance2: 1e-12,
                    initialState: {},
                    ...options,
                },
            };
        }
        default:
            throw new Error(`Unknown fitting algorithm`);
    }
}
